### PR TITLE
Optimize GzipMaxInputIT in order to reduce the JVM memory consumption

### PR DESCRIPTION
**Pull request by jcarranzan**
### Summary

Refactors `GzipMaxInputIT` to reduce memory usage, avoiding spikes in memory consumption.

I compared prints on our Jenkins jobs in the previous one GzipMaxInputIT and observed a rapid spike in memory consumption, but currently, these changes are almost negligible.

This approach addresses the memory issues observed in Podman Jenkins jobs related to Keycloak container failures.

Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [x] Refactoring
- [ ] Backport
- [ ] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [ ] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

**Comment by rsvoboda** [comment]
Do you have jvisualvm or JDK Mission Control screenshots of memory usage for case before and after? So the impact of your changes can be more clear.

**Comment by jcarranzan** [comment]
From my jcarranz-baremetal-ts-jvm-podman-nightly #33 job execution (you look there for more details) : 
```
12:04:01 [After sendPayloadBelowMaxInputLimit Test]
12:04:01 === JVM Memory Info ===
12:04:01 Free Memory: 554 MB
12:04:01 Total Memory: 1002 MB
12:04:01 Max Memory: 1924 MB
12:04:01 === System Memory Info ===
12:04:01 Total Physical Memory: 7695 MB
12:04:01 Free Physical Memory: 1359 MB
12:04:01 --------------------------------------------------
12:04:01 11:04:01,234 INFO  ## Running test GzipMaxInputIT.sendInvalidContent()
12:04:01 [After sendInvalidContent Test]
12:04:01 === JVM Memory Info ===
12:04:01 Free Memory: 260 MB
12:04:01 Total Memory: 1002 MB
12:04:01 Max Memory: 1924 MB
12:04:01 === System Memory Info ===
12:04:01 Total Physical Memory: 7695 MB
12:04:01 Free Physical Memory: 1257 MB
12:04:01 --------------------------------------------------
12:04:01 11:04:01,394 INFO  ## Running test GzipMaxInputIT.sendMaximumAllowedPayload()
12:04:02 [After sendMaximumAllowedPayload Test]
12:04:02 === JVM Memory Info ===
12:04:02 Free Memory: 347 MB
12:04:02 Total Memory: 1365 MB
12:04:02 Max Memory: 1924 MB
12:04:02 === System Memory Info ===
12:04:02 Total Physical Memory: 7695 MB
12:04:02 Free Physical Memory: 794 MB
12:04:02 --------------------------------------------------
12:04:02 11:04:02,397 INFO  ## Running test GzipMaxInputIT.sendMoreThanMaximumAllowedPayload()
12:04:03 [After sendMoreThanMaximumAllowedPayload Test]
12:04:03 === JVM Memory Info ===
12:04:03 Free Memory: 245 MB
12:04:03 Total Memory: 1566 MB
12:04:03 Max Memory: 1924 MB
12:04:03 === System Memory Info ===
12:04:03 Total Physical Memory: 7695 MB
12:04:03 Free Physical Memory: 364 MB
12:04:03 --------------------------------------------------
12:04:03 11:04:03,973 INFO  ## Running test GzipMaxInputIT.sendZeroBytesPayload()
12:04:03 [After sendZeroBytesPayload Test]
12:04:03 === JVM Memory Info ===
12:04:03 Free Memory: 244 MB
12:04:03 Total Memory: 1868 MB
12:04:03 Max Memory: 1924 MB
12:04:03 === System Memory Info ===
12:04:03 Total Physical Memory: 7695 MB
12:04:03 Free Physical Memory: 130 MB
```
Currently with my changes (you can see in the same job #33 execution) : 
```
10:12:24 === JVM Memory Info ===
10:12:24 Free Memory: 393 MB
10:12:24 Total Memory: 1029 MB
10:12:24 Max Memory: 1924 MB
10:12:24 === System Memory Info ===
10:12:24 Total Physical Memory: 7695 MB
10:12:24 Free Physical Memory: 1443 MB
10:12:24 --------------------------------------------------
10:12:24 [After sending invalid content]
10:12:24 === JVM Memory Info ===
10:12:24 Free Memory: 390 MB
10:12:24 Total Memory: 1029 MB
10:12:24 Max Memory: 1924 MB
10:12:24 === System Memory Info ===
10:12:24 Total Physical Memory: 7695 MB
10:12:24 Free Physical Memory: 1425 MB
10:12:24 --------------------------------------------------
10:12:24 09:12:24,456 INFO  ## Running test GzipMaxInputIT.sendInvalidContent()
10:12:24 [Before sending maximum allowed payload]
10:12:24 === JVM Memory Info ===
10:12:24 Free Memory: 390 MB
10:12:24 Total Memory: 1029 MB
10:12:24 Max Memory: 1924 MB
10:12:24 === System Memory Info ===
10:12:24 Total Physical Memory: 7695 MB
10:12:24 Free Physical Memory: 1425 MB
10:12:24 --------------------------------------------------
10:12:24 [After sending invalid content]
10:12:24 === JVM Memory Info ===
10:12:24 Free Memory: 390 MB
10:12:24 Total Memory: 1029 MB
10:12:24 Max Memory: 1924 MB
10:12:24 === System Memory Info ===
10:12:24 Total Physical Memory: 7695 MB
10:12:24 Free Physical Memory: 1425 MB
10:12:24 --------------------------------------------------
10:12:24 09:12:24,473 INFO  ## Running test GzipMaxInputIT.sendMaximumAllowedPayload()
10:12:24 [Before sending maximum allowed payload]
10:12:24 === JVM Memory Info ===
10:12:24 Free Memory: 390 MB
10:12:24 Total Memory: 1029 MB
10:12:24 Max Memory: 1924 MB
10:12:24 === System Memory Info ===
10:12:24 Total Physical Memory: 7695 MB
10:12:24 Free Physical Memory: 1423 MB
10:12:24 --------------------------------------------------
10:12:25 [After sending invalid content]
10:12:25 === JVM Memory Info ===
10:12:25 Free Memory: 389 MB
10:12:25 Total Memory: 1029 MB
10:12:25 Max Memory: 1924 MB
10:12:25 === System Memory Info ===
10:12:25 Total Physical Memory: 7695 MB
10:12:25 Free Physical Memory: 1068 MB
10:12:25 --------------------------------------------------
10:12:25 09:12:25,423 INFO  ## Running test GzipMaxInputIT.sendMoreThanMaximumAllowedPayload()
10:12:25 [Before sending maximum allowed payload]
10:12:25 === JVM Memory Info ===
10:12:25 Free Memory: 389 MB
10:12:25 Total Memory: 1029 MB
10:12:25 Max Memory: 1924 MB
10:12:25 === System Memory Info ===
10:12:25 Total Physical Memory: 7695 MB
10:12:25 Free Physical Memory: 1068 MB
10:12:25 --------------------------------------------------
10:12:26 [After sending invalid content]
10:12:26 === JVM Memory Info ===
10:12:26 Free Memory: 387 MB
10:12:26 Total Memory: 1029 MB
10:12:26 Max Memory: 1924 MB
10:12:26 === System Memory Info ===
10:12:26 Total Physical Memory: 7695 MB
10:12:26 Free Physical Memory: 1062 MB
10:12:26 --------------------------------------------------
10:12:26 09:12:26,897 INFO  ## Running test GzipMaxInputIT.sendZeroBytesPayload()
10:12:26 [Before sending maximum allowed payload]
10:12:26 === JVM Memory Info ===
10:12:26 Free Memory: 387 MB
10:12:26 Total Memory: 1029 MB
10:12:26 Max Memory: 1924 MB
10:12:26 === System Memory Info ===
10:12:26 Total Physical Memory: 7695 MB
10:12:26 Free Physical Memory: 1062 MB
10:12:26 --------------------------------------------------
10:12:26 [After sending invalid content]
10:12:26 === JVM Memory Info ===
10:12:26 Free Memory: 387 MB
10:12:26 Total Memory: 1029 MB
10:12:26 Max Memory: 1924 MB
10:12:26 === System Memory Info ===
10:12:26 Total Physical Memory: 7695 MB
10:12:26 Free Physical Memory: 1062 MB
10:12:26 --------------------------------------------------
10:12:28 09:12:28,962 INFO  [app] Service stopped (Quarkus JVM mode)
10:12:28 [INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time
```

**Review by michalvavrik — COMMENTED** [review]
I think that changing byte array to stream is completely fine and so are changes. Let's wait for green CI before we continue.

Now, the PR description is completely different matter. How I understand code changes in this PR:

- Response used to be passes as byte array to RestAssured, now it is stream
- Memory-wise. improvement is that at no point, you have big byte array on side of integration test
- that is your code, you hope that RestAssure implemetns this efficiently (more efficiently with stream then with array)

So that is "client side". And now "server side":

- `io.quarkus.ts.http.advanced.GzipResource` accepts byte array, so Quarkus side will process data up to `quarkus.resteasy.gzip.max-input`

TL;DR; I can see that difference for array `OVER_LIMIT_PAYLOAD` if RestAssured configures underlying client right, though I think we would need to test that.

And now this PR description:

> Refactors GzipMaxInputIT to reduce memory usage, avoiding spikes in memory consumption.
> I compared prints on our Jenkins jobs in the previous one GzipMaxInputIT and observed a rapid spike in memory consumption, but currently, these changes are almost negligible.

I must be honest, I can't asses it without https://github.com/quarkus-qe/quarkus-test-suite/pull/2170#issuecomment-2461952823, but I trust the data you posted https://github.com/quarkus-qe/quarkus-test-suite/pull/2170#issuecomment-2461972123, so OK.

> This approach addresses the memory issues observed in Podman Jenkins jobs related to Keycloak container failures.

And this is whole reason why I am writing this long comment. You improvements improve `GzipMaxInputIT` memory consumption and memory issue had `HttpAdvancedIT`. When GZIP finishes, memory should be freed? So are you saying that next test is started before that happens?

This is memory state you provided before without changes in this PR:

```
12:04:03 [After sendZeroBytesPayload Test]
12:04:03 === JVM Memory Info ===
12:04:03 Free Memory: 244 MB
12:04:03 Total Memory: 1868 MB
12:04:03 Max Memory: 1924 MB
12:04:03 === System Memory Info ===
12:04:03 Total Physical Memory: 7695 MB
12:04:03 Free Physical Memory: 130 MB
```
and this is after this PR:

```
10:12:26 [After sending invalid content]
10:12:26 === JVM Memory Info ===
10:12:26 Free Memory: 387 MB
10:12:26 Total Memory: 1029 MB
10:12:26 Max Memory: 1924 MB
10:12:26 === System Memory Info ===
10:12:26 Total Physical Memory: 7695 MB
10:12:26 Free Physical Memory: 1062 MB
```

Questions:

- How many times have you tried it? I think memory is funny thing, I'd like to know that this is consistent improvement
- Where is remaining 6633 MB?

I am ready to approve though, I trust you did your research.

**Comment by michalvavrik** [comment]
I wonder why `many-extensions` appeared between changed files, please try to rebase n current main.

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/GzipMaxInputIT.java [hunk 1]
```diff
@@ -20,11 +21,12 @@
...
-    final byte[] zero_bytes = new byte[0];
     final String invalid_value = "";
-    final byte[] SMALL_PAYLOAD = new byte[512];
-    final byte[] LIMIT_PAYLOAD = new byte[100 * 1024 * 1024];
-    final byte[] OVER_LIMIT_PAYLOAD = new byte[200 * 1024 * 1024];
+    final long SMALL_PAYLOAD = 512;
+    final long LIMIT_PAYLOAD = 100 * 1024 * 1024;
+    final long OVER_LIMIT_PAYLOAD = 200 * 1024 * 1024;
+
+    private final byte[] buffer = new byte[4096];
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/GzipMaxInputIT.java:29]: I wrote that note when I thought I was reviewing Jakarta REST resource so I'll close this one (not the bad idea though anyway).

**Comment by jcarranzan** [comment]
> I wonder why `many-extensions` appeared between changed files, please try to rebase n current main.

I don't see it, are u checking with a git command?

**Comment by michalvavrik** [comment]
> > I wonder why `many-extensions` appeared between changed files, please try to rebase n current main.
> 
> I don't see it, are u checking with a git command?

CI failed

**Review by michalvavrik — APPROVED** [review]
Right, so after chagnes before `GzipMaxInputIT` is run, we start with

```
10:12:24 Total Physical Memory: 7695 MB
10:12:24 Free Physical Memory: 1443 MB
```

and we end with 

```
10:12:26 Total Physical Memory: 7695 MB
10:12:26 Free Physical Memory: 1062 MB
```

which is better than before. I think there are still open questions, I suggested some, but I don't mind this PR to go in. Thank you for investigating this matter.

**Comment by jcarranzan** [comment]
@michalvavrik related to your questions:

    How many times have you tried it? I think memory is funny thing, I'd like to know that this is consistent improvement.
    I am running again some Jenkins jobs, but this morning and yesterday locally I also did some runs. I agree wh you regarding to get this as a consistent improvement.
    Where is remaining 6633 MB? While I included also the printout of both memories(JVM and System Memory) this PR is just about JVM memory usage within our tests. That 6633MB is system memory, probably used by other processes and so on.
